crates/ruff_server/src/session/settings.rs

```diff
@@ -1,25 +1,284 @@
-            .unwrap_or_default();
-        let code_action_edit_resolution = code_action_settings
-            .and_then(|code_action_settings| code_action_settings.resolve_support.as_ref())
-            .is_some_and(|resolve_support| resolve_support.properties.contains(&"edit".into()));
+use std::path::Path;
+
+use lsp_types::Url;
+use rustc_hash::FxHashMap;
+use serde::Deserialize;
+
+#[cfg(test)]
+mod tests;
+
+/// Built from the initialization options sent by the client.
+/// If workspace settings were specified, `workspace_settings` will be populated.
+/// Otherwise, it will be empty.
+#[derive(Debug)]
+pub(crate) struct SettingsController {
+    global_settings: UserSettings,
+    workspace_settings: FxHashMap<Url, UserSettings>,
+}
+
+/// Resolved user settings for a specific document. These settings are meant to be
+/// used directly by the server, and are *not* a 1:1 representation with how the client
+/// sends them.
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[allow(dead_code)]
+pub(crate) struct ResolvedUserSettings {
+    fix_all: FixAll,
+    organize_imports: OrganizeImports,
+    lint_enable: LintEnable,
+    run: RunWhen,
+    disable_rule_comment_enable: CodeActionEnable,
+    fix_violation_enable: CodeActionEnable,
+    log_level: LogLevel,
+}
+
+/// This is a direct representation of the user settings schema sent by the client.
+#[derive(Debug, Deserialize, Default)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct UserSettings {
+    fix_all: Option<FixAll>,
+    organize_imports: Option<OrganizeImports>,
+    lint: Option<Lint>,
+    code_action: Option<CodeAction>,
+    log_level: Option<LogLevel>,
+}
+
+/// This is a direct representation of the workspace settings schema,
+/// which inherits the schema of [`UserSettings`] and adds extra fields
+/// to describe the workspace it applies to.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct WorkspaceSettings {
+    #[serde(flatten)]
+    user_settings: UserSettings,
+    workspace: Url,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct FixAll(bool);
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct OrganizeImports(bool);
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct Lint {
+    enable: Option<LintEnable>,
+    run: Option<RunWhen>,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct LintEnable(bool);
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+enum RunWhen {
```

Comment:
Maybe `RunTrigger`?

We could also move the `Default` implementation here from down below.

Reply:
We might actually remove this setting for the time being (see [Micha's comment](https://github.com/astral-sh/ruff/pull/10764#pullrequestreview-1979073417))